### DISCUSSION TITLE
Adding lots of functions

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#22351] [Q&A General] [answered]
I'd like to write a class that creates lots (thousands?) of PiecewiseLinear Functions based on some x-y data. What's a good way to create all those functions? General User Object? Action? Is there an object in MOOSE that does something similar?

**Comment by GiudGiud** [comment, marked as answer]
Hello

An Action is likely the way to go.
There are numerous examples of creating functions in actions. 

Guillaume

**Reply by GiudGiud** [reply]
For reading the data for creating the functions, you may consider the DelimitedFileReader btw

**Reply by jessecarterMOOSE** [reply]
I did use an action but ended up rolling my own file  reader. I used MooseUtils to parse the lines because DelimitedFileReader (from what I can tell) only handles floats and the function names are in the file as strings.

**Comment by friedmud** [comment]
Based on your description: It sounds like you may also want to look into the `PiecewiseMultilinear` Function: https://mooseframework.inl.gov/moose/source/functions/PiecewiseMultilinear.html

**Reply by friedmud** [reply]
For generating the data files you also want to look at the CSV combiner: https://mooseframework.inl.gov/python/combine_csv.html
